### ISSUE TITLE
Mapped channels missing from m3u

### ISSUE THREAD

**Issue by Nanganator** [Threadfin/Threadfin#446]
**Describe the bug**
Similar to #375 (I assume as they left no data in the bug). I have channels from a playlist mapped, but in the m3u several of them are missing. What I notice is that the channels which are missing have the same the bit in red beside x-ID as some others. All of them have unique channel name and x-ID.

**Server (please complete the following information):**
- OS: Ubuntu 22.04.3 LTS (Threadfin running in docker from fyb3roptik/threadfin)
- Arch: Intel 64 Bit
- Threadfin Version and Build: 1.2 (15)

Channels with same writing in red. In the below Channel 2 and 7 are missing from the m3u. The others are showing.
![image](https://github.com/user-attachments/assets/98aad037-8640-4227-ac77-ec87a8318621)
![image](https://github.com/user-attachments/assets/c70cbb98-f205-43d2-b2c6-c8f083ddc6ea)
![image](https://github.com/user-attachments/assets/e7ff8b37-a99b-4a32-98e4-48cb4f96b807)
![image](https://github.com/user-attachments/assets/81560b15-2eff-49c3-bbfc-d720cf93e799)

Other channels from the same playlist have unique red writing:
![image](https://github.com/user-attachments/assets/0b39eba7-4056-4c50-876b-cb6540c487cb)

Removing the play list and readding id did nothing. When adding the channels back into the mapping they have the same red writing and don't show in the m3u.

**Comment by Nanganator** [comment]
OK, after a bit of playing around I realise that red writing is the tvg-id coming from the playlist.

What I did notice is if I add all the channels at once (add the playlist, go to mapping, bulk edit, enable all, save) they all show. But the moment I start making any changes (i.e. Correct capitalization in a channel name) channels start disappearing.

**Comment by Fyb3roptik** [comment]
Duplicate of #439

**Comment by Nanganator** [comment]
This is different from #439. I can see the channels in the mapping section. They are just not in the threadfin.m3u. #439 is channels not appearing in the mapping section.

**Comment by wanderleyteixeira** [comment]
Would you mind creating a new release and I can validate if it really fixes #439

**Comment by joshjryan** [comment]
I think this issue was caused by [this commit](https://github.com/Threadfin/Threadfin/commit/b862cbdf5bd0123ff0edd12a06a011d3aad88cea).  This change made it so that `tvg-id` values must be unique, but I don't think they should be. I think some channels might use the same EPG data, so I think duplicating `tvg-id` should be okay.

@Fyb3roptik do recall what that commit was intending to fix?

**Comment by Nanganator** [comment]
Seems to certainly be the tvg-id duplication.
I grabbed a local copy of my suppliers m3u, removed all the duplicate tvg-id's except one. Removed the previous URL from my supplier and added the local m3u instead. All of the channels except 1 of the duplicates showed in the threafdin.m3u perfectly. Edited the m3u again and gave the last remaining duplicate tvg-id a unique id. Had to update and save the playlist twice to get the update to show. Then all channels appeared in the threafdin.m3u perfectly.

Seems tvg-id duplication causes them to be excluded from the threadfin.m3u, despite appearing in mapping perfectly and being able to be enabled and edited fine.

**Comment by joshjryan** [comment]
@Nanganator does your m3u have another ID field?

I don't want to just undo @Fyb3roptik's change, so I was thinking I could submit a PR that checks for for a non-tvg-id ID field, and if that exists, use it. Then if there isn't a non-tvg-id field, fall back to the current behavior that uses the tvg-id as the unique ID.

**Comment by Nanganator** [comment]
This is all the data for each stream I see in the providers m3u :

#EXTINF:-1 tvg-id="F505ANZ" tvg-name="channel 3" tvg-logo="" group-title="Austria Entertainment",Channel 3
#EXTGRP:Austria Entertainment
https://streamurl.

The tvg-id is the only thing I updated on my local version of this m3u and then all channels were then in the threadfin.m3u

**Comment by joshjryan** [comment]
xteve had some documentation about Unique ID fields [here](https://github.com/xteve-project/xTeVe-Documentation/blob/master/en/configuration.md). 

A whole bunch of years ago, I pulled a python script from someone in the xteve discord, which generates unique IDs for channels in your m3u....I just found a reference to it [here](https://emby.media/community/index.php?/topic/64414-using-xteve-with-emby/&do=findComment&comment=924938).

This might be something that you want to add, if the PR I'm about to submit gets approved.

**Comment by joshjryan** [comment]
@Fyb3roptik I just took a stab at adding support for unique IDs in the PR I just linked here.  It's taking me a bit to wrap my head around how some of this works, so I figured I'd check in what I have and see if you have any input.

**Comment by Nanganator** [comment]
I noticed more peculiar behaviour with this tvg-id duplication issue. I have 2 channels in my supplier m3u "Top Sports 1" and "Top Sports 2". Threadfin shows them both as separate channels in the mapping section and I have only enabled "Top Sports 1". However, because they both have the same tvg-id, the url in the threadfin m3u points to the stream of "Top Sports 2", even though the channel is not enabled.

I've worked around this by disabling the second channel in my supplier side so it is not in the playlist for Threadfin to get confused with.

Not sure if it makes a difference but the channel 1 appears first in the suppliers m3u and then channel 2 is later. I assume it is working top to bottom and using the last match of tvg-id.
